### ISSUE TITLE
Ruby 3.0.0 App Server Crashing

### ISSUE THREAD

**Issue by OpConTech** [sqreen/ruby-agent#5]
When trying to use sqreen ruby agent 1.22.1 with Ruby 3.0.0 and Rails 6.1.0 the app server (rails server) crashes. Looks like Ruby 2.x uses two arguments while Ruby 3.x is only looking for one argument.

Logs:
Jan 07 14:41:56   => Booting Puma
Jan 07 14:41:56   => Rails 6.1.0 application starting in production 
Jan 07 14:41:56   => Run `bin/rails server --help` for more startup options
Jan 07 14:41:58   Build succeeded
Jan 07 14:42:01   Exiting
**Jan 07 14:42:01   ruby/3.0.0/gems/puma-5.1.1/lib/rack/handler/puma.rb:62:in `run': wrong number of arguments (given 2, expected 1) (ArgumentError)
Jan 07 14:42:01   ruby/3.0.0/gems/sqreen-1.22.1/lib/sqreen/graft/hook.rb:257:in `block (3 levels) in wrapper'**
Jan 07 14:42:01   ruby/3.0.0/gems/sqreen-1.22.1/lib/sqreen/graft/call.rb:129:in `ignore'
Jan 07 14:42:01   ruby/3.0.0/gems/sqreen-1.22.1/lib/sqreen/graft/hook.rb:255:in `block (2 levels) in wrapper'
Jan 07 14:42:01   ruby/3.0.0/gems/sqreen-1.22.1/lib/sqreen/graft/call.rb:149:in `measure'
Jan 07 14:42:01   ruby/3.0.0/gems/sqreen-1.22.1/lib/sqreen/graft/hook.rb:182:in `block in wrapper'
Jan 07 14:42:01   ruby/3.0.0/gems/rack-2.2.3/lib/rack/server.rb:327:in `start'

**Comment by lloeki** [comment]
Hey @OpConTech thanks for the feedback.

> Looks like Ruby 2.x uses two arguments while Ruby 3.x is only looking for one argument.

It's a bit (a lot actually) more complex than that:

Here are the relevant lines:

- https://github.com/puma/puma/blob/v5.1.1/lib/rack/handler/puma.rb#L62
- https://github.com/rack/rack/blob/2.2.3/lib/rack/server.rb#L327

And the matching calls:

```
# rack
server.run(wrapped_app, **options, &block)
# puma
def self.run(app, **options)
```

Both signatures match, irrespective to the Ruby version.

The problem comes form [this Ruby 3.0 breaking change](https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/).

The corner cases listed in the doc are relatively easy to fix for an app that targets only one Ruby version, hard for a library that targets multiple ruby versions while maintaining existing calls compatible, and hair-pulling for instrumentation tools that inject themselves into existing apps and libs (so both defines a method that receives arguments from the caller -and- then calls back the original method, attempting to preserve the call behaviour).

The current version of the agent uses `*args`, which when used by itself usually propagates kwargs properly. Why keep the `*args`? because:

- the current version still supports Ruby 1.9.3, which does not support the `**kwargs` syntax.
- there is a slight behaviour difference when using `*args, **kwargs` between ruby 2.6+2.7 and 3.0 depending on the last argument value when it is a hash (string vs symbol keys), this is worked around by using the ruby2_keywords thing, but it cannot be used by instrumentors as is.

The next version of the agent will drop support for 1.9.3, and we will be able to move to `**kwargs`, but we must also ensure that we don't break things on these hash keys corner cases.

**Comment by OpConTech** [comment]
@lloeki thanks for helping me so quickly.

Just to summarize, there is no work around for my Ruby 3 right now but the next version of the Ruby sqreen agent will support Ruby 3 and while also keeping support for Ruby 2.x.  

So sounds like I will wait for the next sqreen agent release. 

Thanks again for the help.

**Comment by lloeki** [comment]
> Just to summarize, there is no work around for my Ruby 3 right now but the next version of the Ruby sqreen agent will support Ruby 3 and while also keeping support for Ruby 2.x.

Untested, but you could try to drop to Puma 4.x which doesn't use kwargs.

**Comment by OpConTech** [comment]
@lloeki  --- Ha! your right. I bumped Puma down from 5.1.1 (latest) to 4.3.6 (last 4.x) and puma app server is able to load the current sqreen agent just fine. More testing to do but Puma v4.3.6 could be a work around.

**Comment by OpConTech** [comment]
@lloeki  -  I noticed that the sqreen gem bumped to 1.23.0 (was 1.22.1).  So tried loading it with latest version of puma web server (5.1.1) but kwargs issue seems to continue.

**Comment by lloeki** [comment]
Thanks for the report. That is expected, 1.23.0 does not fix it: I have an internal agent version that fixes the root cause but making `**kwargs` work consistently through instrumentation all the way down to 2.0 (which we support) is ridiculously hard.

**Comment by lloeki** [comment]
Fixed in 1.23.1.
